File: jar/src/main/java/jline/api/mam/Mamap2m_fit_trace.java
/**
 * @file Markovian Arrival MAP with Marked arrivals trace-based fitting
 *
 * Fits MAMAP(2,m) processes from empirical trace data with inter-arrival times and class labels.
 * Essential for data-driven modeling of real multiclass arrival processes.
 *
 * @since LINE 3.0
 */
package jline.api.mam;

import java.util.HashMap;
import java.util.Map;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class Mamap2m_fit_trace {
    private Mamap2m_fit_trace() {}

    /**
     * Fits a MAMAP(2,m) to trace data.
     *
     * @param interArrivalTimes Array of inter-arrival times
     * @param classLabels Array of class labels for each arrival
     * @return Fitted MAMAP(2,m) model
     */
    public static MatrixCell mamap2m_fit_trace(double[] interArrivalTimes, int[] classLabels) {
        if (interArrivalTimes.length != classLabels.length) {
            throw new IllegalArgumentException("Inter-arrival times and class labels must have same length");
        }

        // Extract basic moments
        double M1 = average(interArrivalTimes);
        double M2 = averageSquared(interArrivalTimes);
        double M3 = averageCubed(interArrivalTimes);

        // Extract class information
        int maxLabel = 0;
        for (int label : classLabels) {
            if (label > maxLabel) maxLabel = label;
        }
        int numClasses = maxLabel + 1;
        int[] classCounts = new int[numClasses];
        for (int label : classLabels) {
            if (label >= 0 && label < numClasses) {
                classCounts[label]++;
            }
        }

        double[] classProbs = new double[numClasses];
        for (int i = 0; i < numClasses; i++) {
            classProbs[i] = (double) classCounts[i] / classLabels.length;
        }
        double[] backwardMoments = computeClassBackwardMoments(interArrivalTimes, classLabels, numClasses);

        // Create proper matrices for the function call
        MatrixCell dummyMap = new MatrixCell(2);
        dummyMap.set(0, new Matrix(2, 2));
        dummyMap.set(1, new Matrix(2, 2));

        double[] F = new double[]{M2}; // Forward moments
        double[] B = backwardMoments;  // Backward moments

        // Returns a 3-tuple whose first element is the MatrixCell.
        return Mamap2m_fit_fb_multiclass.mamap2m_fit_fb_multiclass(dummyMap, classProbs, F, B).mmap;
    }

    /**
     * Compute backward moments for each class
     */
    private static double[] computeClassBackwardMoments(double[] interArrivalTimes, int[] classLabels, int numClasses) {
        double[] backwardMoments = new double[numClasses];
        Map<Integer, Double> lastArrivalTime = new HashMap<Integer, Double>();
        double currentTime = 0.0;

        for (int i = 0; i < interArrivalTimes.length; i++) {
            currentTime += interArrivalTimes[i];
            int currentClass = classLabels[i];

            if (currentClass >= 0 && currentClass < numClasses) {
                if (lastArrivalTime.containsKey(currentClass)) {
                    double backwardTime = currentTime - lastArrivalTime.get(currentClass);
                    backwardMoments[currentClass] = backwardTime;
                } else {
                    backwardMoments[currentClass] = currentTime;
                }
                lastArrivalTime.put(currentClass, currentTime);
            }
        }

        // Fill in missing values with global mean
        double globalMean = average(interArrivalTimes);
        for (int i = 0; i < backwardMoments.length; i++) {
            if (backwardMoments[i] == 0.0) {
                backwardMoments[i] = globalMean;
            }
        }

        return backwardMoments;
    }

    private static double average(double[] arr) {
        double s = 0.0;
        for (double v : arr) s += v;
        return s / arr.length;
    }

    private static double averageSquared(double[] arr) {
        double s = 0.0;
        for (double v : arr) s += v * v;
        return s / arr.length;
    }

    private static double averageCubed(double[] arr) {
        double s = 0.0;
        for (double v : arr) s += v * v * v;
        return s / arr.length;
    }
}

File: jar/src/main/kotlin/jline/api/mam/Mamap2m_fit_trace.kt
/**
 * @file Markovian Arrival MAP with Marked arrivals trace-based fitting
 * 
 * Fits MAMAP(2,m) processes from empirical trace data with inter-arrival times and class labels.
 * Essential for data-driven modeling of real multiclass arrival processes.
 * 
 * @since LINE 3.0
 */
package jline.api.mam

import jline.util.matrix.MatrixCell
import jline.util.matrix.Matrix

/**
 * Fits a MAMAP(2,m) to trace data.
 *
 * @param interArrivalTimes Array of inter-arrival times
 * @param classLabels Array of class labels for each arrival
 * @return Fitted MAMAP(2,m) model
 */
fun mamap2m_fit_trace(interArrivalTimes: DoubleArray, classLabels: IntArray): MatrixCell {
    require(interArrivalTimes.size == classLabels.size) {
        "Inter-arrival times and class labels must have same length"
    }
    
    // Extract basic moments
    val M1 = interArrivalTimes.average()
    val M2 = interArrivalTimes.map { it * it }.average()
    val M3 = interArrivalTimes.map { it * it * it }.average()
    
    // Extract class information
    val numClasses = (classLabels.maxOrNull() ?: 0) + 1
    val classCounts = IntArray(numClasses)
    for (label in classLabels) {
        if (label >= 0 && label < numClasses) {
            classCounts[label]++
        }
    }
    
    val classProbs = classCounts.map { it.toDouble() / classLabels.size }.toDoubleArray()
    val backwardMoments = computeClassBackwardMoments(interArrivalTimes, classLabels, numClasses)
    
    // Create proper matrices for the function call
    val dummyMap = MatrixCell(2)
    dummyMap[0] = Matrix(2, 2)
    dummyMap[1] = Matrix(2, 2)
    
    val F = doubleArrayOf(M2) // Forward moments
    val B = backwardMoments   // Backward moments
    
    val result = mamap2m_fit_fb_multiclass(dummyMap, classProbs, F, B)
    return result.first // Return just the MatrixCell part of the Triple
}

/**
 * Compute backward moments for each class
 */
private fun computeClassBackwardMoments(
    interArrivalTimes: DoubleArray,
    classLabels: IntArray,
    numClasses: Int
): DoubleArray {
    
    val backwardMoments = DoubleArray(numClasses)
    val lastArrivalTime = mutableMapOf<Int, Double>()
    var currentTime = 0.0
    
    for (i in interArrivalTimes.indices) {
        currentTime += interArrivalTimes[i]
        val currentClass = classLabels[i]
        
        if (currentClass >= 0 && currentClass < numClasses) {
            if (lastArrivalTime.containsKey(currentClass)) {
                val backwardTime = currentTime - lastArrivalTime[currentClass]!!
                backwardMoments[currentClass] = backwardTime
            } else {
                backwardMoments[currentClass] = currentTime
            }
            lastArrivalTime[currentClass] = currentTime
        }
    }
    
    // Fill in missing values with global mean
    val globalMean = interArrivalTimes.average()
    for (i in backwardMoments.indices) {
        if (backwardMoments[i] == 0.0) {
            backwardMoments[i] = globalMean
        }
    }
    
    return backwardMoments
}
/**
 * MAMAP 2m fit trace algorithms
 */
@Suppress("unused")
class Mamap2mFitTraceAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}